### PR TITLE
Change the discrete trajectory iterator to be a CRTP class to avoid the odd implicit constructor

### LINKED ISSUE #707: CRTP within CRTP
This would make Forkable::Iterator easier to use (no need to artificially create DiscreteTrajectory::Iterator).  And we must go deeper.

### PR DESCRIPTION
**Pull request by pleroy**
Fix #707.

### CODE REVIEW

File: physics/forkable.hpp [hunk 1]
```diff
@@ -15,6 +15,23 @@ using geometry::Instant;
 
 namespace physics {
 
+// Forkable and ForkableIterator both use CRTP to achieve static polymorphism on
+// the parameters and return type of the member functions: we want them to
+// return Tr4jectory and It3rator, not Forkable and ForkableIterator, so that
+// the clients don't have to down_cast or construct object of subclasses.
```
Comment (eggrobin) [inline physics/forkable.hpp]: grammars.

File: physics/forkable.hpp [hunk 2]
```diff
@@ -22,21 +39,73 @@ namespace physics {
...
+  ForkableIterator() = default;
+
+  bool operator==(It3rator const& right) const;
+  bool operator!=(It3rator const& right) const;
+
+  It3rator& operator++();
+  It3rator& operator--();
+
+  // Returns the point in the timeline that is denoted by this iterator.
+  TimelineConstIterator current() const;
```
Comment (eggrobin) [inline physics/forkable.hpp]: might be private?

File: physics/forkable.hpp [hunk 3]
```diff
@@ -22,21 +39,73 @@ namespace physics {
...
+
+ protected:
+  // The API that must be implemented by subclasses.
+  // Must return |this| of the proper type
```
Comment (eggrobin) [inline physics/forkable.hpp]: .

File: physics/forkable.hpp [hunk 4]
```diff
@@ -53,71 +122,22 @@ class Forkable {
...
-    Iterator& operator++();
-    Iterator& operator--();
-
-    // Returns the point in the timeline that is denoted by this iterator.
-    TimelineConstIterator current() const;
-
-   protected:
-    // Returns the (most forked) trajectory to which this iterator applies.
-    not_null<Tr4jectory const*> trajectory() const;
+  using Iterator = It3rator;
```
Comment (eggrobin) [inline physics/forkable.hpp]: nuit grave

File: physics/forkable_test.cpp [hunk 5]
```diff
@@ -18,21 +18,38 @@ namespace physics {
...
+
   FakeTrajectory() = default;
 
   void push_back(Instant const& time);
 
-  using Forkable<FakeTrajectory>::NewFork;
-  using Forkable<FakeTrajectory>::DeleteAllForksAfter;
-  using Forkable<FakeTrajectory>::DeleteAllForksBefore;
+  using Forkable<FakeTrajectory,
+                 internal::FakeTrajectoryIterator>::NewFork;
```
Comment (eggrobin) [inline physics/forkable_test.cpp]: Iterator

File: physics/forkable_test.cpp [hunk 6]
```diff
@@ -49,15 +66,29 @@ class FakeTrajectory : public Forkable<FakeTrajectory> {
...
+namespace internal {
+
+not_null<FakeTrajectoryIterator*> FakeTrajectoryIterator::that() {
+  return this;
+}
+
+not_null<FakeTrajectoryIterator const*> FakeTrajectoryIterator::that() const {
+  return this;
+}
+
```
Comment (eggrobin) [inline physics/forkable_test.cpp]: order bodies

File: physics/discrete_trajectory.hpp [hunk 7]
```diff
@@ -36,26 +38,31 @@ struct ForkableTraits<DiscreteTrajectory<Frame>> {
...
 
 template<typename Frame>
-class DiscreteTrajectory : public Forkable<DiscreteTrajectory<Frame>> {
+class DiscreteTrajectoryIterator
+    : public ForkableIterator<DiscreteTrajectory<Frame>,
+                              DiscreteTrajectoryIterator<Frame>> {
+ public:
+  Instant const& time() const;
+  DegreesOfFreedom<Frame> const& degrees_of_freedom() const;
+ protected:
```
Comment (eggrobin) [inline physics/discrete_trajectory.hpp]: blank line before

File: physics/forkable.hpp (hunk 1 above)
Comment (pleroy) [inline physics/forkable.hpp]: Done.

File: physics/forkable.hpp (hunk 2 above)
Comment (pleroy) [inline physics/forkable.hpp]: No, but protected.

File: physics/forkable.hpp (hunk 3 above)
Comment (pleroy) [inline physics/forkable.hpp]: Done.

File: physics/forkable.hpp (hunk 4 above)
Comment (pleroy) [inline physics/forkable.hpp]: Done.

File: physics/forkable_test.cpp (hunk 5 above)
Comment (pleroy) [inline physics/forkable_test.cpp]: Done.

File: physics/forkable_test.cpp (hunk 6 above)
Comment (pleroy) [inline physics/forkable_test.cpp]: Done.

File: physics/discrete_trajectory.hpp (hunk 7 above)
Comment (pleroy) [inline physics/discrete_trajectory.hpp]: Done.

**Pull request merged**